### PR TITLE
Proper authenticator option support

### PR DESCRIPTION
**Pull request by iltar**
This PR removes the need for the hostname option, as this is not a mandatory parameter, but merely something you can add to the accountName. It also adds backwards compatibility for older versions of the authenticator using the old issuer notation. Aditionally, the new signature allows you to define the size of the generated image, making it easier to plug into a templating engine for example.

```php
// old was restricted to
$authenticator->getUrl('foo', 'foobar.org', '3DHTQX4GCRKHGS55CJ', 'FooBar');
// otpauth://totp/foo@foobar.org?secret=3DHTQX4GCRKHGS55CJ&issuer=FooBar

// new notation would be
GoogleQrUrl::generate('foo@foobar.org', '3DHTQX4GCRKHGS55CJ', 'FooBar')
// otpauth://totp/FooBar:foo@foobar.org?secret=3DHTQX4GCRKHGS55CJ&issuer=FooBar

// additionally, you can now properly create the label to not include @hostname
GoogleQrUrl::generate('John Doe', '3DHTQX4GCRKHGS55CJ', 'FooBar')
// otpauth://totp/FooBar:John Doe?secret=3DHTQX4GCRKHGS55CJ&issuer=FooBar
```
_The generated URLs in the tests are all valid and tested with the google authenticator app itself._

Most notable changes:
 * Deprecated `GoogleAuthenticator::getUrl()` in favor of `GoogleQrUrl::generate`.
 * Added `GoogleQrUrl::generate()` which contains more validation checks and is
   more compatible with what Google provides.

## Changelog

```markdown
### Added
- Added `GoogleQrUrl::generate` which contains more validation checks and is
   more compatible with what Google provides
### Changed
- Deprecated `GoogleAuthenticator::getUrl()` in favor of `GoogleQrUrl::generate()`
```

### CODE REVIEW

File: UPGRADE-2.x.md [hunk 1]
```diff
@@ -1,2 +1,9 @@
 UPGRADE 2.x
 ===========
+
+### GoogleAuthenticator
+
+ * Deprecated `GoogleAuthenticator::getUrl()` in favor of `GoogleQrUrl::generate`.
```
Comment (OskarStark) [inline UPGRADE-2.x.md]: generate() (with brackets)

Comment (OskarStark) [inline UPGRADE-2.x.md]: headline: deprecated xYz and body, how someone could avoid the deprecate including the signature, I would say

File: UPGRADE-2.x.md [hunk 2]
```diff
@@ -1,2 +1,9 @@
...
+ 
+ * Added `GoogleQrUrl::generate` which contains more validation checks and is
```
Comment (OskarStark) [inline UPGRADE-2.x.md]: IMO we don't need this in the UPGRADE file

File: src/ExceptionInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * file that was distributed with this source code.
+ */
+namespace Google\Authenticator;
+
+/**
+ * Indicates the interface is coming from this package.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+interface ExceptionInterface
```
Comment (OskarStark) [inline src/ExceptionInterface.php]: why do we need this class?

File: src/GoogleAuthenticator.php [hunk 4]
```diff
@@ -119,17 +119,23 @@ public function getCode($secret, $time = null): string
      */
     public function getUrl($user, $hostname, $secret): string
```
Comment (OskarStark) [inline src/GoogleAuthenticator.php]: `@deprecated` tag is missing

File: src/GoogleQrUrl.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        $otpauthString = rawurlencode(sprintf($otpauthString, $label, $secret, $issuer));
+
+        return sprintf('https://chart.googleapis.com/chart?chs=%1$dx%1$d&chld=M|0&cht=qr&chl=%2$s', $size, $otpauthString);
+    }
+
+    /**
+     * Private by design.
+     */
+    private function __construct()
```
Comment (OskarStark) [inline src/GoogleQrUrl.php]: do we need this?

**Comment by OskarStark** [comment]
Great work @iltar 👍

File: src/GoogleQrUrl.php (hunk 5 above)
Comment (iltar) [inline src/GoogleQrUrl.php]: In theory not, but it's just to ensure that people do not instantiate the class (should be a static class)

File: src/ExceptionInterface.php (hunk 3 above)
Comment (iltar) [inline src/ExceptionInterface.php]: Oh, I forgot to catch the exceptions based on the new request class, I still have to add those! 

This interface was supposed to be for that purpose, but now that I think of it, maybe not really needed

Comment (iltar) [inline src/ExceptionInterface.php]: Actually, there's no need to catch those exceptions as before that, the generation would've failed and there's no way to get around it other than removing the invalid values manually :thinking:

File: src/GoogleAuthenticator.php (hunk 4 above)
Comment (iltar) [inline src/GoogleAuthenticator.php]: Added with the same message as the trigger error

File: tests/GoogleQrUrlTest.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        $this->assertEquals(
+            'https://chart.googleapis.com/chart?chs=400x400&chld=M|0&cht=qr&chl=otpauth%3A%2F%2Ftotp%2Ffoo%40foobar.org%3Fsecret%3D3DHTQX4GCRKHGS55CJ',
+            GoogleQrUrl::generate('foo@foobar.org', '3DHTQX4GCRKHGS55CJ', null, 400)
+        );
+    }
+
+    /**
+     * @expectedException \Google\Authenticator\RuntimeException
+     * @expectedExceptionMessage The issuer name may not contain a double colon (:) and may not be an empty string. Given "".
+     */
```
Comment (greg0ire) [inline tests/GoogleQrUrlTest.php]: We usually prefer methods over annotations for this but here your test is one line long so it's not a big deal IMO

Comment (iltar) [inline tests/GoogleQrUrlTest.php]: I don't mind either way, but I do know that as 2 phpunit versions are used, there might be some method incompatibilities regarding the exception expectations

File: src/RuntimeException.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Contains runtime exception templates.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class RuntimeException extends \RuntimeException
+{
+    public static function InvalidAccountName(string $accountName): self
+    {
+        return new self(sprintf('The account name may not contain a double colon (:) and may not be an empty string. Given "%s".', $accountName));
```
Comment (greg0ire) [inline src/RuntimeException.php]: Please linebreak this (not the message, just the sprintf arguments), and do the same for methods below that exceed 120 chars

File: tests/GoogleQrUrlTest.php (hunk 6 above)
Comment (greg0ire) [inline tests/GoogleQrUrlTest.php]: There are, and we usually implement a polyfill for that. Not worth it in this case IMO.

File: src/GoogleQrUrl.php [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ */
+final class GoogleQrUrl
+{
+    /**
+     * NEXT_MAJOR: make a private const once 7.1 is allowed.
+     *
+     * @inernal
+     * @private
+     */
+    const URL_TEMPLATE = 'https://chart.googleapis.com/chart?chs=%s&chld=M|0&cht=qr&chl=%s';
```
Comment (greg0ire) [inline src/GoogleQrUrl.php]: This seems unused

Comment (iltar) [inline src/GoogleQrUrl.php]: Correct! nice catch

**Comment by greg0ire** [comment]
Thanks @iltar !

**Pull request merged**